apps/emqx_dashboard/src/emqx_dashboard_swagger.erl

```diff
@@ -0,0 +1,226 @@
+        responses => Responses},
+        lists:usort(Refs1 ++ Refs2 ++ Refs3)
+    }.
+
+parameters(Params) ->
+    lists:foldl(fun({Name, Schema}, {Acc, RefsAcc}) ->
+        #{in := In, type := HoconType} = Schema,
+        {Type, Refs} = hocon_type_to_spec(HoconType),
+        Spec0 = maps:without([type], Schema),
+        Spec1 = Spec0#{schema => Type, name => Name},
+        Spec2 = add_required_to_path(In, Spec1),
+        {[Spec2 | Acc], Refs ++ RefsAcc}
+                end, {[], []}, Params).
+
+add_required_to_path(path, Spec) -> Spec#{required => true};
+add_required_to_path(_, Spec) -> Spec.
+
+%% only object
+requestBody(Request) when is_map(Request) ->
+    case parse_object(Request) of
+        {[], Refs} -> {[], Refs};
+        {Props, Refs} ->
+            {#{<<"content">> =>
+            #{<<"application/json">> => #{<<"schema">> => Props}}}, Refs}
+    end.
+
+
+responses(Responses) ->
+    maps:fold(fun(Status, Hocon, {Acc, RefsAcc}) ->
+        if is_map(Hocon) ->
+            case maps:find(type, Hocon) of
+                {ok, HoconType} ->
+                    {Spec, Refs} = hocon_type_to_spec(HoconType),
+                    Init = maps:without([type], Hocon),
+                    Content = #{<<"application/json">> => #{<<"schema">> => Spec}},
+                    {Acc#{integer_to_binary(Status) => Init#{<<"content">> => Content}}, Refs ++ RefsAcc};
+                error -> %% object
+                    {Props, Refs} = parse_object(Hocon),
+                    Object = #{<<"content">> => #{<<"application/json">> => #{<<"schema">> => Props}}},
+                    {Acc#{integer_to_binary(Status) => Object}, Refs ++ RefsAcc}
+            end;
+            is_binary(Hocon) ->
+                {Acc#{integer_to_binary(Status) => #{description => Hocon}}, RefsAcc}
+        end
+              end, {#{}, []}, Responses).
+
+components(Refs) ->
+    maps:fold(fun(K, V, Acc) -> [#{K => V} | Acc] end, [], components(Refs, #{}, [])).
+
+components([], SpecAcc, []) -> SpecAcc;
+components([], SpecAcc, SubRefAcc) -> components(SubRefAcc, SpecAcc, []);
+components([{Module, Field} | Refs], SpecAcc, SubRefsAcc) ->
+    Props = apply(Module, fields, [Field]),
+    case is_list(Props) of
+        true -> %% object
+            {Object, SubRefs} = parse_object(maps:from_list(Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => Object},
+            components(Refs, NewSpecAcc, SubRefs ++ SubRefsAcc);
+        false -> %% schema
+            Init = maps:without([type], Props),
+            {Spec, SubRef} = hocon_type_to_spec(maps:get(type, Props)),
+            NewSpecAcc = SpecAcc#{?REF(Module, Field) => maps:merge(Spec, Init)},
+            components(Refs, NewSpecAcc, SubRef ++ SubRefsAcc)
+    end.
+
+hocon_type_to_spec(?R_REF(Module, Field)) ->
+    {#{<<"$ref">> => ?TO_COMPONENTS(Module, Field)}, [{Module, Field}]};
+hocon_type_to_spec(Type) when ?IS_TYPEREFL(Type) ->
+    Match = #{
```

Comment:
a proper way is to allow typerefl export an API to retrieve the `name` part of the type metadata in this pattern `{'$type_refl', #{name := Name}}`

for duration, percent, file etc, we should use:
https://github.com/k32/typerefl/blob/master/src/typerefl.erl#L103-L113

Reply:
I try to use `alias/2`, but `alias/2` makes it more complex and difficult to use.

1. we define alias.
```erlang
-module(emqx_swagger).
duration_ms() ->
    typerefl:alias("duration_ms", emqx_schema:duration_ms()).

file() ->
    typerefl:alias("file", emqx_schema:file()).

percent() ->
    typerefl:alias("percent", emqx_schema:percent()).

ip_port() ->
    typerefl:alias("ip_port", emqx_schema:ip_port()).

range(Min, Max) ->
    typerefl:alias(["range", Min, Max], typerefl:range(Min, Max)).
...
hocon_schema_to_spec(Type) when ?IS_TYPEREFL(Type) ->
    Spec =
        case typerefl:name(Type) of
            "boolean()" -> #{type => boolean, example => true};
            "binary()" -> #{type => string, example => <<"binary example">>};
            "float()" -> #{type => number, example => 3.14159};
            "integer()" -> #{type => integer, example => 100};
            "number()" -> #{type => number, example => 42};
            ["string","(",[],")"] -> #{type => string, example => <<"string example">>};
            "atom()" -> #{type => string, example => atom};
            ["duration", "(", [], ")"] -> #{type => string, example => "12m"};
            ["duration_s", "(", [], ")"] -> #{type => string, example => "1h"};
            ["duration_ms", "(", [], ")"] -> #{type => string, example => "32s"};
            ["file", "(", [], ")"] -> #{type => string, example => "/path/to/file"};
            ["ip_port", "(", [], ")"] -> #{type => string, example => "127.0.0.1:80"};
            ["percent", "(", [], ")"] -> #{type => number, example => "12%"};
            [["range", Min, Max], "(", [], ")"] ->
                #{type => integer, example => Min, minimum => Min, maximum => Max}
        end,
    {Spec, []}.

```
2. use alias.
```erlang
-module(xxx_api).
-include_lib("typerefl/include/types.hrl").
-import(emqx_swagger, [duration_s/0, ip_port/0, range/2]).
```
This causes 2 problems:
- types.hrl and emqx_swagger both  import range/2, which can make using it look very confusing, (in fact, only typer.hrl's range/2 is working, we must call emqx_swagger:range/2)
- the hocon schema don't use those alias, If we reference hocon schema fields (in config side), their name never match the alias. so I use the emqx_schema definition directly on the spec side for consistency of use.

So I suggest to keep it this way, maybe you have a better way?

Reply:
I mean `typerefl:alias("string",  emqx_schema:duration_ms())`

Reply:
and there should be no `[]` inside the `()`, i.e. it's `string()`, not `string([])`

Reply:
I see what you mean, and sent a PR here to fix: https://github.com/k32/typerefl/pull/29